src/stats/manager_impl.go

```diff
@@ -0,0 +1,156 @@
+package stats
+
+import (
+	pb_struct "github.com/envoyproxy/go-control-plane/envoy/extensions/common/ratelimit/v3"
+	"github.com/envoyproxy/ratelimit/src/settings"
+	gostats "github.com/lyft/gostats"
+	logger "github.com/sirupsen/logrus"
+)
+
+func NewStatManager(store gostats.Store, s settings.Settings) *ManagerImpl {
+	logger.Infof("Initializing Stat Manager with detailed metrics %t", s.DetailedMetrics)
+	serviceScope := store.ScopeWithTags("ratelimit", s.ExtraTags).Scope("service")
+	return &ManagerImpl{
+		store:             store,
+		rlStatsScope:      serviceScope.Scope("rate_limit"),
+		legacyStatsScope:  serviceScope.Scope("call.should_rate_limit_legacy"),
+		serviceStatsScope: serviceScope,
+		detailed:          s.DetailedMetrics,
+	}
+}
+
+type ManagerImpl struct {
+	store             gostats.Store
+	rlStatsScope      gostats.Scope
+	legacyStatsScope  gostats.Scope
+	serviceStatsScope gostats.Scope
+	detailed          bool
+}
+
+func (this *ManagerImpl) GetStatsStore() gostats.Store {
+	return this.store
+}
+
+func (this *ManagerImpl) AddTotalHits(u uint64, rlStats RateLimitStats, key string) {
+	rlStats.TotalHits.Add(u)
+	if this.detailed && key != rlStats.Key {
+		stat := this.getDescriptorStat(key)
+		stat.TotalHits.Add(u)
+	}
+}
+
+func (this *ManagerImpl) AddOverLimit(u uint64, rlStats RateLimitStats, key string) {
+	rlStats.OverLimit.Add(u)
+	if this.detailed && key != rlStats.Key {
+		stat := this.getDescriptorStat(key)
+		stat.OverLimit.Add(u)
+	}
+}
+
+func (this *ManagerImpl) AddNearLimit(u uint64, rlStats RateLimitStats, key string) {
+	rlStats.NearLimit.Add(u)
+	if this.detailed && key != rlStats.Key {
+		stat := this.getDescriptorStat(key)
+		stat.NearLimit.Add(u)
+	}
+}
+
+func (this *ManagerImpl) AddOverLimitWithLocalCache(u uint64, rlStats RateLimitStats, key string) {
+	rlStats.OverLimitWithLocalCache.Add(u)
+	if this.detailed && key != rlStats.Key {
+		stat := this.getDescriptorStat(key)
+		stat.OverLimitWithLocalCache.Add(u)
+	}
+}
+
+//todo: consider adding a RateLimitStats cache
+//todo: consider adding descriptor fields parameter to allow configuration of descriptor entries for which metrics will be emited.
+func (this *ManagerImpl) getDescriptorStat(key string) RateLimitStats {
```

Comment:
I understand that you're using TittleCase for the methods' names. In that case, shouldn't this be "GetDescriptorStat"?

Reply:
Golang function/method names are capitalization sensitive.
Methods that start with capital letter such as `GetDescriptorStat` are called exported methods (similar to java's `public` modifier)
Methods that start with lower-case letter such as `getDescriptorStat` are called unexported methods (similar to java's `protected` modifier)

Reply:
Thank you very much for the comment, now I know a little bit more about Go :)